chain-crypto/src/algorithms/legacy_daedalus.rs

```diff
@@ -0,0 +1,138 @@
+use crate::key::{
+  AsymmetricKey, AsymmetricPublicKey, PublicKeyError, SecretKeyError, SecretKeySizeStatic,
+};
+use crate::sign::{SignatureError, SigningAlgorithm, Verification, VerificationAlgorithm};
+
+use super::ed25519 as ei;
+use cryptoxide::ed25519;
+use ed25519_bip32::{XPrv, XPub, XPRV_SIZE, XPUB_SIZE};
+use rand_core::{CryptoRng, RngCore};
+
+pub const CHAIN_CODE_SIZE: usize = 32;
+
+/// Legacy Daedalus algorithm
+pub struct LegacyDaedalus;
+
+#[derive(Clone)]
+pub struct LegacyPriv([u8; XPRV_SIZE]);
+
+impl AsRef<[u8]> for LegacyPriv {
+    fn as_ref(&self) -> &[u8] {
+        &self.0[..]
+    }
+}
+
+impl LegacyPriv {
+  pub fn from_xprv(xprv: &XPrv) -> Self {
+    let mut buf = [0; XPRV_SIZE];
+    buf[0..XPRV_SIZE].clone_from_slice(xprv.as_ref());
+    LegacyPriv(buf)
+  }
+
+  pub fn inner_key(&self) -> [u8; ed25519::PRIVATE_KEY_LENGTH] {
+    let mut buf = [0; ed25519::PRIVATE_KEY_LENGTH];
+    buf[0..ed25519::PRIVATE_KEY_LENGTH].clone_from_slice(&self.0.as_ref()[0..ed25519::PRIVATE_KEY_LENGTH]);
+    buf
+  }
+
+  pub fn chaincode(&self) -> [u8; CHAIN_CODE_SIZE] {
+    let mut buf = [0; CHAIN_CODE_SIZE];
+    buf[0..CHAIN_CODE_SIZE].clone_from_slice(&self.0.as_ref()[ed25519::PRIVATE_KEY_LENGTH..XPRV_SIZE]);
+    buf
+  }
+}
+
+impl AsymmetricPublicKey for LegacyDaedalus {
+  type Public = XPub;
+  const PUBLIC_BECH32_HRP: &'static str = "legacy_xpub";
+  const PUBLIC_KEY_SIZE: usize = XPUB_SIZE;
+  fn public_from_binary(data: &[u8]) -> Result<Self::Public, PublicKeyError> {
+      let xpub = XPub::from_slice(data)?;
+      Ok(xpub)
+  }
+}
+
+impl AsymmetricKey for LegacyDaedalus {
+  type Secret = LegacyPriv;
+  type PubAlg = LegacyDaedalus;
+
+  const SECRET_BECH32_HRP: &'static str = "legacy_xprv";
+
+  fn generate<T: RngCore + CryptoRng>(mut _rng: T) -> Self::Secret {
+      panic!("generation not implemented for legacy Daedalus wallets")
```

Comment:
The panic here causes the test at the bottom of this file to fail. Not sure if there's a good way to provide an arbitrary instance only for when the test is called.